api/rest/client/lbclient.go

```diff
@@ -12,74 +13,87 @@ import (
 	peer "github.com/libp2p/go-libp2p-peer"
 )
 
-// loadBalancingClient a client to interact with IPFS Cluster APIs.
+// loadBalancingClient is a client to interact with IPFS Cluster APIs.
 // It balances the load by distributing it among peers.
 type loadBalancingClient struct {
 	strategy LBStrategy
+	retries  int
 }
 
-// LBStrategy is a strategy to load balance request among clients
+// LBStrategy is a strategy to load balance request among clients.
 type LBStrategy interface {
-	Next(count int, call func(Client) error) error
+	Next(count int) Client
+	SetClients(clients []Client)
 }
 
 // RoundRobin is a load balancing strategy that would use clients in a sequence.
 type RoundRobin struct {
 	clients []Client
 	length  int
-	retries int
-	counter chan int
+	counter *int32
 }
 
 // NewRoundRobin would return an LBStrategy that load balances requests by using
 // clients in sequence.
-func NewRoundRobin(cfgs []*Config, retries int) (LBStrategy, error) {
-	var clients []Client
-	for _, cfg := range cfgs {
-		defaultClient, err := NewDefaultClient(cfg)
-		if err != nil {
-			return nil, err
-		}
-		clients = append(clients, defaultClient)
-	}
-
+func NewRoundRobin() LBStrategy {
 	rand.Seed(time.Now().UnixNano())
 
-	counter := make(chan int)
-	counter <- 0
-	return &RoundRobin{clients: clients, length: len(clients), retries: retries, counter: counter}, nil
+	var counter *int32
+	atomic.StoreInt32(counter, 0)
+	return &RoundRobin{counter: counter}
 }
 
-// Next return the next client to be used
-func (r *RoundRobin) Next(count int, call func(Client) error) error {
-	i := <-r.counter
-	r.counter <- (i + 1) % r.length
-	err := call(r.clients[i])
-	apiErr, ok := err.(*api.Error)
-	if !ok {
-		logger.Error("could not cast error into api.Error")
-		return err
-	}
+// Next return the next client to be used.
+func (r *RoundRobin) Next(count int) Client {
+	i := atomic.LoadInt32(r.counter)
+	atomic.StoreInt32(r.counter, (i+1)%int32(r.length))
```

Comment:
This will return the same client to two concurrent callers because the steps between load and atomic are not "atomic". You can use atomic add and do the module at the end. I don't think overflow is a problem (but check), as long as you use uints..

Reply:
> I don't think overflow is a problem (but check), as long as you use uints

true, starts from 0 after max https://play.golang.org/p/XKh3hYS6zPO